<a href="https://colab.research.google.com/github/standroidbeta/DS-Unit-1-Sprint-2-Data-Wrangling/blob/master/Sprintchallenge2/DS_Unit_1_Sprint_Challenge_2_Data_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 2

## Data Wrangling

In this Sprint Challenge you will use data from [Gapminder](https://www.gapminder.org/about-gapminder/), a Swedish non-profit co-founded by Hans Rosling. "Gapminder produces free teaching resources making the world understandable based on reliable statistics."
- [Cell phones (total), by country and year](https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--cell_phones_total--by--geo--time.csv)
- [Population (total), by country and year](https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--population_total--by--geo--time.csv)
- [Geo country codes](https://github.com/open-numbers/ddf--gapminder--systema_globalis/blob/master/ddf--entities--geo--country.csv)

These two links have everything you need to successfully complete the Sprint Challenge!
- [Pandas documentation: Working with Text Data](https://pandas.pydata.org/pandas-docs/stable/text.html]) (one question)
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf) (everything else)

## Part 0. Load data

You don't need to add or change anything here. Just run this cell and it loads the data for you, into three dataframes.

In [0]:
import pandas as pd

cell_phones = pd.read_csv('https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--cell_phones_total--by--geo--time.csv')

population = pd.read_csv('https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--population_total--by--geo--time.csv')

geo_country_codes = (pd.read_csv('https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--entities--geo--country.csv')
                       .rename(columns={'country': 'geo', 'name': 'country'}))

In [2]:
cell_phones

geo  time  cell_phones_total
0     abw  1960                0.0
1     abw  1965                0.0
2     abw  1970                0.0
3     abw  1975                0.0
4     abw  1976                0.0
5     abw  1977                0.0
6     abw  1978                0.0
7     abw  1979                0.0
8     abw  1980                0.0
9     abw  1981                0.0
10    abw  1982                0.0
11    abw  1983                0.0
12    abw  1984                0.0
13    abw  1985                0.0
14    abw  1986                0.0
15    abw  1987                0.0
16    abw  1988                0.0
17    abw  1989                0.0
18    abw  1990                0.0
19    abw  1991                0.0
20    abw  1992               20.0
21    abw  1995             1718.0
22    abw  1996             3000.0
23    abw  1997             3402.0
24    abw  1998             5380.0
25    abw  1999            12000.0
26    abw  2000            15000.0
27    abw  2001            53000.0
28    abw  2002            61800.0
29    abw  2003            69952.0
...   ...   ...                ...
9185  zwe  1988                0.0
9186  zwe  1989                0.0
9187  zwe  1990                0.0
9188  zwe  1991                0.0
9189  zwe  1992                0.0
9190  zwe  1993                0.0
9191  zwe  1994                0.0
9192  zwe  1995                0.0
9193  zwe  1996                0.0
9194  zwe  1997             5734.0
9195  zwe  1998            19000.0
9196  zwe  1999           174000.0
9197  zwe  2000           266441.0
9198  zwe  2001           314002.0
9199  zwe  2002           338779.0
9200  zwe  2003           363651.0
9201  zwe  2004           425745.0
9202  zwe  2005           647110.0
9203  zwe  2006           849146.0
9204  zwe  2007          1225654.0
9205  zwe  2008          1654721.0
9206  zwe  2009          3991000.0
9207  zwe  2010          7700000.0
9208  zwe  2011          9200000.0
9209  zwe  2012         12613935.0
9210  zwe  2013         13633167.0
9211  zwe  2014         11798652.0
9212  zwe  2015         12757410.0
9213  zwe  2016         12878926.0
9214  zwe  2017         14092104.0

[9215 rows x 3 columns]

In [3]:
cell_phones.shape

(9215, 3)

In [20]:
cell_phones['geo'].value_counts()

cyp            46
fra            46
ukr            46
syr            46
tur            46
nic            46
dnk            46
com            46
pak            46
moz            46
mwi            46
ind            46
irq            46
gnb            46
deu            46
zwe            46
btn            46
mar            46
bdi            46
pry            46
phl            46
ltu            46
hti            46
slv            46
jpn            46
lka            46
bra            46
cri            46
nor            46
mlt            46
               ..
esp            43
sgp            42
sle            42
abw            41
cym            41
kna            41
atg            39
qat            38
are            36
mhl            36
nru            35
gum            33
asm            30
vir            28
mnp            28
tca            28
isr            28
sau            28
smr            27
vgb            26
pse            22
lie            21
tls            16
srb            14
mne       

In [24]:
cell_phones.isnull().sum()

geo                  0
time                 0
cell_phones_total    0
dtype: int64

In [4]:
population

geo  time  population_total
0      afg  1800           3280000
1      afg  1801           3280000
2      afg  1802           3280000
3      afg  1803           3280000
4      afg  1804           3280000
5      afg  1805           3280000
6      afg  1806           3280000
7      afg  1807           3280000
8      afg  1808           3280000
9      afg  1809           3280000
10     afg  1810           3280000
11     afg  1811           3280779
12     afg  1812           3282342
13     afg  1813           3284692
14     afg  1814           3287834
15     afg  1815           3291770
16     afg  1816           3296506
17     afg  1817           3302044
18     afg  1818           3308390
19     afg  1819           3315547
20     afg  1820           3323519
21     afg  1821           3332311
22     afg  1822           3341926
23     afg  1823           3352368
24     afg  1824           3363642
25     afg  1825           3375751
26     afg  1826           3388701
27     afg  1827           3402494
28     afg  1828           3417136
29     afg  1829           3432630
...    ...   ...               ...
59267  zwe  2071          36422461
59268  zwe  2072          36674084
59269  zwe  2073          36918818
59270  zwe  2074          37156394
59271  zwe  2075          37386604
59272  zwe  2076          37609317
59273  zwe  2077          37824533
59274  zwe  2078          38032210
59275  zwe  2079          38232406
59276  zwe  2080          38425112
59277  zwe  2081          38610322
59278  zwe  2082          38787965
59279  zwe  2083          38957958
59280  zwe  2084          39120196
59281  zwe  2085          39274605
59282  zwe  2086          39421192
59283  zwe  2087          39559970
59284  zwe  2088          39691002
59285  zwe  2089          39814374
59286  zwe  2090          39930170
59287  zwe  2091          40038459
59288  zwe  2092          40139287
59289  zwe  2093          40232710
59290  zwe  2094          40318796
59291  zwe  2095          40397615
59292  zwe  2096          40469281
59293  zwe  2097          40533896
59294  zwe  2098          40591626
59295  zwe  2099          40642628
59296  zwe  2100          40687107

[59297 rows x 3 columns]

In [5]:
population.shape

(59297, 3)

In [22]:
population['geo'].value_counts()

cyp    301
isl    301
eri    301
srb    301
bhs    301
svn    301
cmr    301
cze    301
zmb    301
nga    301
cri    301
aze    301
ben    301
ukr    301
pak    301
cod    301
hos    301
rou    301
tkm    301
bhr    301
geo    301
mli    301
ssd    301
arm    301
tha    301
mlt    301
prt    301
usa    301
yem    301
sle    301
      ... 
jam    301
tza    301
uga    301
kir    301
mmr    301
swe    301
mus    301
hun    301
sur    301
bra    301
mkd    301
mdg    301
tuv    301
dnk    301
wsm    301
aut    301
hrv    301
mne    301
mda    301
civ    301
lso    301
col    301
lux    301
stp    301
gab    301
bel    301
ecu    301
hkg    301
bol    301
plw    301
Name: geo, Length: 197, dtype: int64

In [23]:
population.isnull().sum()

geo                 0
time                0
population_total    0
dtype: int64

In [6]:
geo_country_codes.head()

geo alt_5                    alternative_1 alternative_2 alternative_3  \
0  abkh   NaN                              NaN           NaN           NaN   
1   abw   NaN                              NaN           NaN           NaN   
2   afg   NaN  Islamic Republic of Afghanistan           NaN           NaN   
3   ago   NaN                              NaN           NaN           NaN   
4   aia   NaN                              NaN           NaN           NaN   

  alternative_4_cdiac arb1 arb2 arb3 arb4         ...           latitude  \
0                 NaN  NaN  NaN  NaN  NaN         ...                NaN   
1               Aruba  NaN  NaN  NaN  NaN         ...           12.50000   
2         Afghanistan  NaN  NaN  NaN  NaN         ...           33.00000   
3              Angola  NaN  NaN  NaN  NaN         ...          -12.50000   
4                 NaN  NaN  NaN  NaN  NaN         ...           18.21667   

  longitude main_religion_2008      country        pandg un_state  \
0       NaN                NaN     Abkhazia          NaN    False   
1 -69.96667          christian        Aruba          NaN    False   
2  66.00000             muslim  Afghanistan  AFGHANISTAN     True   
3  18.50000          christian       Angola       ANGOLA     True   
4 -63.05000          christian     Anguilla          NaN    False   

  unicode_region_subtag  upper_case_name world_4region        world_6region  
0                   NaN              NaN        europe  europe_central_asia  
1                    AW            ARUBA      americas              america  
2                    AF      AFGHANISTAN          asia           south_asia  
3                    AO           ANGOLA        africa   sub_saharan_africa  
4                    AI         ANGUILLA      americas              america  

[5 rows x 33 columns]

## Part 1. Join data

First, join the `cell_phones` and `population` dataframes (with an inner join on `geo` and `time`).

The resulting dataframe's shape should be: (8590, 4)

In [0]:
cell_phones_filter = cell_phones.drop(cell_phones[cell_phones['cell_phones_total'] < 1].index)

In [8]:
cell_phones_filter.shape

(5157, 3)

In [9]:
cell_phones_filter.head()

geo  time  cell_phones_total
20  abw  1992               20.0
21  abw  1995             1718.0
22  abw  1996             3000.0
23  abw  1997             3402.0
24  abw  1998             5380.0

In [0]:
population_filter = population.drop(population[population['time'] < 1992].index)

In [12]:
population_filter.head()

geo  time  population_total
192  afg  1992          13981231
193  afg  1993          15095099
194  afg  1994          16172719
195  afg  1995          17099541
196  afg  1996          17822884

In [0]:
population_filter = population_filter.drop(columns=['population_total'])

In [14]:
population_filter.shape

(21473, 2)

In [0]:
cp_merged = pd.merge(cell_phones_filter, population_filter, how='inner', on='geo')

In [16]:
cp_merged.shape

(527015, 4)

In [17]:
cp_merged.head()

geo  time_x  cell_phones_total  time_y
0  afg    2002            25000.0    1992
1  afg    2002            25000.0    1993
2  afg    2002            25000.0    1994
3  afg    2002            25000.0    1995
4  afg    2002            25000.0    1996

In [0]:
# cp_merged = cp_merged.drop(cp_merged[cp_merged['time_x'] >= 1992].index)

In [0]:
# cp_merged = cp_merged.drop(cp_merged[cp_merged['time_y'] >= 1992].index)

In [0]:
# cp_merged.shape

## Part 2. Make features

Calculate the number of cell phones per person, and add this column onto your dataframe.

(You've calculated correctly if you get 1.220 cell phones per person in the United States in 2017.)

Modify the `geo` column to make the geo codes uppercase instead of lowercase.

## Part 3. Process data

Use the describe function, to describe your dataframe's numeric columns, and then its non-numeric columns.

(You'll see the time period ranges from 1960 to 2017, and there are 195 unique countries represented.)

In 2017, what were the top 5 countries with the most cell phones total?

Your list of countries should have these totals:

| country | cell phones total |
|:-------:|:-----------------:|
|    ?    |     1,474,097,000 |
|    ?    |     1,168,902,277 |
|    ?    |       458,923,202 |
|    ?    |       395,881,000 |
|    ?    |       236,488,548 |



In [0]:
# This optional code formats float numbers with comma separators
pd.options.display.float_format = '{:,}'.format

2017 was the first year that China had more cell phones than people.

What was the first year that the USA had more cell phones than people?

## Part 4. Reshape data

Create a pivot table:
- Columns: Years 2007—2017
- Rows: China, India, United States, Indonesia, Brazil (order doesn't matter)
- Values: Cell Phones Total

The table's shape should be: (5, 11)

#### OPTIONAL BONUS QUESTION!

Sort these 5 countries, by biggest increase in cell phones from 2007 to 2017.

Which country had 935,282,277 more cell phones in 2017 versus 2007?

If you have the time and curiosity, what other questions can you ask and answer with this data?